### PR TITLE
 regionliveness: probe on replica errors for liveness on rbr system tables

### LINKED ISSUE #119236: roachtest: acceptance/multitenant-multiregion failed
roachtest.acceptance/multitenant-multiregion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14006082?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14006082?buildTab=artifacts#/acceptance/multitenant-multiregion) on master @ [cc6ca026319024800395293b0fb18f05dd8eb50e](https://github.com/cockroachdb/cockroach/commits/cc6ca026319024800395293b0fb18f05dd8eb50e):

```
(assertions.go:333).Fail: 
	Error Trace:	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/multitenant.go:237
	            				github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/acceptance.go:150
	            				main/pkg/cmd/roachtest/test_runner.go:1129
	            				src/runtime/asm_amd64.s:1650
	Error:      	Received unexpected error:
	            	no time out detect because of dead region.
	            	(1) secondary error attachment
	            	  | pq: operation "count-leases-region" timed out after 15.002s (given timeout 15s): count-leases: failed to read query result: query execution canceled
	            	  | (1) pq: operation "count-leases-region" timed out after 15.002s (given timeout 15s): count-leases: failed to read query result: query execution canceled
	            	  | Error types: (1) *pq.Error
	            	Wraps: (2) assertion failure
	            	Wraps: (3) attached stack trace
	            	  -- stack trace:
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runAcceptanceMultitenantMultiRegion.func3
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/multitenant.go:218
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runAcceptanceMultitenantMultiRegion.Group.GoCtx.func6
	            	  | 	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168
	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	            	  | 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75
	            	  | runtime.goexit
	            	  | 	src/runtime/asm_amd64.s:1650
	            	Wraps: (4) no time out detect because of dead region.
	            	Error types: (1) *secondary.withSecondaryError (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
	Test:       	acceptance/multitenant-multiregion
	Messages:   	Waited for go routines, expected no error.
(require.go:1376).NoErrorf: FailNow called
test artifacts and logs in: /artifacts/acceptance/multitenant-multiregion/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_fs=ext4</code>
 - <code>ROACHTEST_localSSD=true</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-14006082/acceptance-multitenant-multiregion/1707979196023/1707979660490)

</p>
</details>
/cc @cockroachdb/multi-tenant
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*acceptance/multitenant-multiregion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-36102

### LINKED ISSUE #121327: roachtest: acceptance/multitenant-multiregion failed
roachtest.acceptance/multitenant-multiregion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14610723?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14610723?buildTab=artifacts#/acceptance/multitenant-multiregion) on release-24.1 @ [557f4e6fe89d683d49082d88ba5adf6660e8f16f](https://github.com/cockroachdb/cockroach/commits/557f4e6fe89d683d49082d88ba5adf6660e8f16f):

```
(test_runner.go:1185).runTest: test timed out (10m0s)
test artifacts and logs in: /artifacts/acceptance/multitenant-multiregion/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_fs=ext4</code>
 - <code>ROACHTEST_localSSD=true</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-14610723/acceptance-multitenant-multiregion/1711691098748/1711691870409)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #119236 roachtest: acceptance/multitenant-multiregion failed [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*acceptance/multitenant-multiregion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-37188

### PR DESCRIPTION
**Pull request by fqazi**
Previously, we probed for region liveness issues only after encountering
a timeout error when accessing regional-by-row tables. This approach
missed retriable replica errors (such as slow proposals), which also
indicate potential availability problems. This patch now treats
retriable replica errors as a trigger for region liveness probes,
improving the speed of issue detection. Additionally, we'll update the
multi-region multi-tenant test to ensure the timely propagation of span
configuration changes.

Fixes: #121327
Fixes: https://github.com/cockroachdb/cockroach/issues/119236

Release note: None

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121485)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @fqazi, and @srosenberg)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSC-qk7Z9zsnt5Dv0Q):*
i may have the wrong understanding -- but i thought these were not actually retried? the errors are just used to detect up/down.

also, some of these sentences might have had some missed wording, so it's a little hard to follow this
___
*[`pkg/cmd/roachtest/tests/multitenant.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSAnMu0YJmTtOjKDaW:-NuSAnMu0YJmTtOjKDaX:bxsr94o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6444545a495a9d6d4e95aeec750329ca3fa8e2c7/pkg/cmd/roachtest/tests/multitenant.go#L162)):*
> ```Go
> 	// propagated, we'll shut down the tenant and wait for them to get
> 	// applied.
> 	tdb, tdbCloser := openDBAndMakeSQLRunner(t, tenants[0].pgURL)
> ```

nit: we have the same exact code in `TestColdStartLatency`. should we make a shared function that does this?

(curious what you think; not sure if it's worth it)
___
*[`pkg/sql/regionliveness/prober.go` line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSB6LZ3Tv3MXSiFFsy:-NuSB6L_-mxjKrrk6XNC:bpzh3rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6444545a495a9d6d4e95aeec750329ca3fa8e2c7/pkg/sql/regionliveness/prober.go#L346)):*
> ```Go
> 	return pgerror.GetPGCode(err) == pgcode.QueryCanceled ||
> 		errors.HasType(err, (*timeutil.TimeoutError)(nil)) ||
> 		startup.IsRetryableReplicaError(err) ||
> ```

i  think it's a bit misleading to use the `IsRetryableReplicaError` function here -- we aren't actually going to retry these right? since this is just used for the purpose of deciding if a region is alive or not, it might be more clear to spell out the errors explicitly here:
```
		errors.HasType(err, (*kvpb.ReplicaUnavailableError)(nil)) ||
		errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) ||
		pgerror.GetPGCode(err) == pgcode.RangeUnavailable
```

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121485)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, and @srosenberg)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSC-qk7Z9zsnt5Dv0Q):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i may have the wrong understanding -- but i thought these were not actually retried? the errors are just used to detect up/down.

also, some of these sentences might have had some missed wording, so it's a little hard to follow this
</blockquote></details>

Sorry, yeah this was really poorly worded (updated this commit message):

regionliveness: probe on replica errors for liveness on rbr system tables

Previously, we probed for region liveness issues only after encountering
a timeout error when accessing regional-by-row tables. This approach
missed retriable replica errors (such as slow proposals), which also
indicate potential availability problems. This patch now treats
retriable replica errors as a trigger for region liveness probes,
improving the speed of issue detection. Additionally, we'll update the
multi-region multi-tenant test to ensure the timely propagation of span
configuration changes.
___
*[`pkg/cmd/roachtest/tests/multitenant.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSAnMu0YJmTtOjKDaW:-NuUYJ-iEHwgoUM4E45i:b-yhhs0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6444545a495a9d6d4e95aeec750329ca3fa8e2c7/pkg/cmd/roachtest/tests/multitenant.go#L162)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we have the same exact code in `TestColdStartLatency`. should we make a shared function that does this?

(curious what you think; not sure if it's worth it)
</blockquote></details>

Good point, I made this code shared between the two of them.
___
*[`pkg/sql/regionliveness/prober.go` line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuSB6LZ3Tv3MXSiFFsy:-NuUXtQf8P8yW1rdVy_U:b-2no0wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6444545a495a9d6d4e95aeec750329ca3fa8e2c7/pkg/sql/regionliveness/prober.go#L346)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i  think it's a bit misleading to use the `IsRetryableReplicaError` function here -- we aren't actually going to retry these right? since this is just used for the purpose of deciding if a region is alive or not, it might be more clear to spell out the errors explicitly here:
```
		errors.HasType(err, (*kvpb.ReplicaUnavailableError)(nil)) ||
		errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) ||
		pgerror.GetPGCode(err) == pgcode.RangeUnavailable
```
</blockquote></details>

Good point, I'll directly propagate the checks here and also dropped 
AmbiguousResultError since that shouldn't be a worry on reads.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121485)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @fqazi, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/util_mutiltregion.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuUlmtnByFIPMLNJm2J:-NuUlmto2ccQi8C1_Cae:b-fbapnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/53132001eb68aec42d6bf6fb3eacc6b4da3d80d7/pkg/cmd/roachtest/tests/util_mutiltregion.go#L1)):*
> ```Go
> package tests
> ```

i think this file would be better in `pkg/testutils`; to avoid weird dependencies

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121485)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, @srosenberg, and @yuzefovich)

___
*[`pkg/cmd/roachtest/tests/util_mutiltregion.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuUlmtnByFIPMLNJm2J:-NuV6HtU9RstFxtCgwZj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/53132001eb68aec42d6bf6fb3eacc6b4da3d80d7/pkg/cmd/roachtest/tests/util_mutiltregion.go#L1)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this file would be better in `pkg/testutils`; to avoid weird dependencies
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
nice fix!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-:-NuV86zw9E7I0TtsxIwB:bhpddey)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @fqazi, @srosenberg, and @yuzefovich)

___
*[`pkg/testutils/sqlutils/util_mutiltregion.go` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121485#-NuV7v8d-bezzrSk-RsP:-NuV7v8d-bezzrSk-RsQ:b-6s7kc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaa782956ca9d9dc3ef7f408e99635ef418fd89a/pkg/testutils/sqlutils/util_mutiltregion.go#L2)):*
> ```Go
> package sqlutils
> ```

nit: this file needs a header for the license (and to pass the linter)

**Comment by fqazi** [comment]
@rafiss TFTR!

bors r+
